data_safe_haven/infrastructure/programs/sre/firewall.py

```diff
@@ -0,0 +1,270 @@
+"""Pulumi component for SRE traffic routing"""
+
+from collections.abc import Mapping
+
+from pulumi import ComponentResource, Input, Output, ResourceOptions
+from pulumi_azure_native import network
+
+from data_safe_haven.infrastructure.common import (
+    get_address_prefixes_from_subnet,
+    get_id_from_subnet,
+)
+from data_safe_haven.types import (
+    FirewallPriorities,
+    PermittedDomains,
+    Ports,
+)
+
+
+class SREFirewallProps:
+    """Properties for SREFirewallComponent"""
+
+    def __init__(
+        self,
+        location: Input[str],
+        resource_group_name: Input[str],
+        route_table_name: Input[str],
+        subnet_apt_proxy_server: Input[network.GetSubnetResult],
+        subnet_firewall: Input[network.GetSubnetResult],
+        subnet_firewall_management: Input[network.GetSubnetResult],
+        subnet_guacamole_containers: Input[network.GetSubnetResult],
+        subnet_identity_containers: Input[network.GetSubnetResult],
+        subnet_user_services_software_repositories: Input[network.GetSubnetResult],
+        subnet_workspaces: Input[network.GetSubnetResult],
+    ) -> None:
+        self.location = location
+        self.resource_group_name = resource_group_name
+        self.route_table_name = route_table_name
+        self.subnet_apt_proxy_server_prefixes = Output.from_input(
+            subnet_apt_proxy_server
+        ).apply(get_address_prefixes_from_subnet)
+        self.subnet_identity_containers_prefixes = Output.from_input(
+            subnet_identity_containers
+        ).apply(get_address_prefixes_from_subnet)
+        self.subnet_firewall_id = Output.from_input(subnet_firewall).apply(
+            get_id_from_subnet
+        )
+        self.subnet_firewall_management_id = Output.from_input(
+            subnet_firewall_management
+        ).apply(get_id_from_subnet)
+        self.subnet_guacamole_containers_prefixes = Output.from_input(
+            subnet_guacamole_containers
+        ).apply(get_address_prefixes_from_subnet)
+        self.subnet_user_services_software_repositories_prefixes = Output.from_input(
+            subnet_user_services_software_repositories
+        ).apply(get_address_prefixes_from_subnet)
+        self.subnet_workspaces_prefixes = Output.from_input(subnet_workspaces).apply(
+            get_address_prefixes_from_subnet
+        )
```

Comment:
I hope we should be able to test this as well, to check all of the transformations do what we want.

Is there any problem using Pulumi `Ouput` and `Input` in tests?

Reply:
Great question. I have a suspicion that they won't work unless we can get the Pulumi unit test framework working (see #1835).